### ISSUE TITLE
Inserting a "rule-hits" entry to response header

### ISSUE THREAD

**Issue by Olament** [SpiderLabs/ModSecurity#2538]
Hi, I am currently working on testing the correctness of CRS and some customized rule sets with ModSecurity. For now, every time I sent a request, I need to go through the debug log to determine if the request hit the targeted rule. I am wondering if it is  possible to make ModSecurity inserting a "rule-hits" entry that contains all of the matched rules in the response header. 

Something like this
```bash
# inbound request
curl --request GET \
  --url http://localhost:8080/ \
  --header 'User-Agent: BAD SCANNER'
```
```
# Response from ModSecurity
HTTP/1.1 403 Forbidden
Transfer-Encoding: chunked
rule-hits: CRS-SCANNER-DETECTION-920170, CRS-BLOCKING-EVALUATION-949110
Date: Tue, 23 Mar 2021 03:23:24 GMT
```
Is it possible to achieve this with a customized rule set? I am thinking add a new transaction variable that store the rule id of the hit rules and then return a customized response with the value of that variable. If this is impossible using the customized rules alone, I think it would be a neat feature to add to the ModSecurity.

**Comment by martinhsv** [comment]
Hi @Olament ,

What setup are you using?  Apache HTTP Server with ModSecurity 2.9.x?  nginx with ModSecurity 3.0.x?

Do you only need this to work for blocked requests?  I.e. you are only looking for the rule information to be in the response if the request never reached the web server (and you do _not_ need the information to be added to real responses from the web server)?

If hypothetically it is impractical to get the information into a response header, can I assume having in the response body would meet your needs?

**Comment by Olament** [comment]
I am using Nginx with ModSecurity 3.x and yes it only need to be worked for blocked request. It does not have to be inside the Response header and having it inside response body is perfectly fine. So it is bascially like a customized 403 page from ModSecurity. 

I did some research myself and the closest thing I can find is the Content Injection (prepend and append element into html body). But unfortunately it was planned but never implemented.

**Comment by hsluoyz** [comment]
@martinhsv , we have made some testcases for the CRS rules. For example, the testcases for rule 920170 are intended to trigger the rule in ModSec by manupulating some fields of the HTTP headers. But when we send the testcase request to ModSec, we can only get response of HTTP status like 200 or 403. If we got 403, we don't know which rule has been triggered and blocked our request.

For a better white box testing, I think ModSec should provide a way (like testing mode) to return the information about what rules have been triggered to the tester. So we can better know if our hand-made testcases work as expected.

Do you know if ModSec has already supported such a mechanism? If not, is there a convenient way to implement it? Currently, we have several proposals as follows.

1. Read the audit or debug logs of the ModSec, as the logs have recorded the triggered rules. But this way requires the tester to access the log files and parse them. This way doesn't support paralell testing, as there's no way to map the log line to the request.

2. Modify ModSec source code. We don't want to do this as it makes it difficult to sync with the upstream in future.

3. Make some custom rules to save the matched result of each rule to states like TX, then use a final rule to return the TX as HTTP response. But I don't know if this way works. And also messing with the rules is not that good.

So our goals are:

1. Get the triggered rules elegantly for each testcase request.
2. Modify the ModSec core code or rules as least as possible. Keep updating ModSec or CRS versions easy.

Is there any suggestion? Thanks!

**Comment by airween** [comment]
@hsluoyz - a couple of years ago I also picked up on this idea. This would be a good help.

For testing, especially for CRS you can use [ftw](https://github.com/coreruleset/ftw) (with Python 2.7).

There are two alternative solutions: [go-ftw](https://github.com/fzipi/go-ftw), which is beta, and [ftwrunner](https://github.com/digitalwave/ftwrunner), which tests only libmodsecurity3 without any HTTP server.

**Comment by hsluoyz** [comment]
@airween actually we are also developing a test framework which is similar to the listed ones. And currently we are stuck at the place: how to get the triggered rules elegantly from the response of the challenging request.

I think it's a valuable feature which even can be integrated into the ModSec core code, as end-2-end testing is very important for WAF. Another obstacle is the new libmodsecurity3 architecture. So if we use Nginx, we will have Nginx + libmodsecurity, libmodsecurity3 cannot add response header like "triggered rules" by itself, it must let Nginx do the job. So we have to modify the code for both Nginx + libmodsecurity, and also for other connectors. I don't know if there's a better direction to do this.

@zimmerle @brenosilva @victorhora

**Comment by martinhsv** [comment]
Hi @hsluoyz ,

If it were me, I would probably pursue one of two options as the simplest that I can think of at the moment -- and they are both variations on two of the three approaches you have already considered:

1. Use your option (1).  To match up audit log entries to the requests that triggered them, I would simply include a test-case identifier as part of the request data that will be in the audit log.  Whether the test case identifier is a unique key in an SQL database or a line number in a file, you could simply include this as an extra path segment in the uri (or elsewhere in the request).

3. Use your option 3.  There is at least one built-in way to return response information purely through ModSecurity rules and without having to do anything fancy with nginx config:  redirects.

The ModSecurity variable RULE contains this information and can be assigned to a tx variable.  Agreed that it is usually undesirable to start modifying a large number of rules in a rule set, but you can do the initial assignment within SecDefaultAction, for example to handle all phase:2 rules:

```
SecDefaultAction "phase:2,<whatever other actions you have>,setvar:'tx.modsec_ruleid=%{RULE.id}'"
```

Then, all you need are two custom rules to run at the end of phase:2.  Something like:

```
SecRule tx.modsec_ruleid "@ge 1" "id:9999998,phase:2,log,setvar:'tx.modsec_ruleid_send=%{tx.modsec_ruleid}'"
SecRule tx:modsec_ruleid_send "@ge 1" "id:9999999,phase:2,log,redirect:http://localhost/403.html?modsec_ruleid=%{tx.modsec_ruleid_send}"
```

**Comment by Olament** [comment]
@martinhsv, Thanks a lot! The second option is what we are looking for. This is a nice workaround.

**Comment by Olament** [comment]
Hi @martinhsv 

Your proposed solutions works well in our testing environment. However, we ran into a problem of displaying correct response status.  Let's say there is a request that trigger a non-critical rule (which means that we expect ModSecurity not to deny the request but log it instead) and there is a request that trigger a critical rule. If we use the redirect method, we cannot tell the difference between those two request since they both have the same response status code "302". To resolve this problem, we set the ModSecurity to AnomalyScoring mode and use the following rule combination

```
SecRule tx.anomaly_score "@ge %{tx.inbound_anomaly_score_threshold}" \
    "id:9999998,phase:2,log,setvar:'tx.response_status=403'"

SecRule tx.anomaly_score "@lt %{tx.inbound_anomaly_score_threshold}" \
    "id:9999997,phase:2,log,setvar:'tx.response_status=200'"

SecRule tx.anomaly_score "@unconditionalMatch" \
    "id:9999999,phase:2,log,\
    redirect:http://localhost/?status=%{tx.response_status}&modsec_ruleid=%{tx.modsec_ruleid_send}"
```

This works but does not seem to be an elegant solution. I am wondering if there is a better way to pass the response status code to our testing client. 

Thanks!
